### ISSUE TITLE
Why are changes not refreshed in the main context?

### ISSUE THREAD

**Issue by antonholmquist** [hyperoslo/Sync#142]
After changes are synced using the temporary background context, shouldn't the updates be refreshed in the main context using `refreshObject(_:mergeChanges:)` before completion?

I figure this is typically what you want. Or is there is a reason not to do this?

Thanks!

**Comment by 3lvis** [comment]
Hi Anton,

Thanks for commenting on this, but I'm afraid I don't follow. [Today the changes get merged to the main context like doing](https://github.com/3lvis/DATAStack/blob/master/Source/DATAStack.swift#L189):

```swift
self.mainContext.mergeChangesFromContextDidSaveNotification(notification)
```

What's the specific use case for using `refreshObject(_:mergeChanges:)`?

**Comment by antonholmquist** [comment]
Hi, thank for the quick reply. I get it, but I wonder when that notification is sent. Maybe not always at the exact same time the context is saved.

We are experiencing that the main context is not always updated by the time the completion block is called when merging changes, which could possibly be due the notification being sent when a small delay. I have not confirmed this though.

Forcing a refresh with `refreshObject` solved this issue for us.

**Comment by 3lvis** [comment]
I guess this is for something that's not using NSFetchedResultsController. Would you mind sharing a snippet of what are you doing?

Would be great to get this fixed :)

The notification gets triggered in the when the `save` method gets called by the temporary `backgroundContext`.

```
func createUser() {
    self.dataStack.performInNewBackgroundContext { backgroundContext in
        let entity = NSEntityDescription.entityForName("User", inManagedObjectContext: backgroundContext)!
        let object = NSManagedObject(entity: entity, insertIntoManagedObjectContext: backgroundContext)
        object.setValue("Background", forKey: "name")
        object.setValue(NSDate(), forKey: "createdDate")
        try! backgroundContext.save() // Here the changes will get merged with the main thread
    }
}
```

**Comment by antonholmquist** [comment]
Yes, exactly. Basically we are trying to fetch the updated object in the completion block using a fetch request. I'm not sure why the merge is not done at the done the completion block runs. Could possible be due the notification being delayed, but I'm not sure.

Will start by investigating and see if I can can track down what is happening. 

Appreciate your support!

**Comment by antonholmquist** [comment]
I'm a little confused. The notification is indeed called like expected, before the completion block.

Not sure this is a framework error. Will continue looking into it..

**Comment by 3lvis** [comment]
I guess you're trying to do something like this?

```swift
func createUser() {
    self.dataStack.performInNewBackgroundContext { backgroundContext in
        let entity = NSEntityDescription.entityForName("User", inManagedObjectContext: backgroundContext)!
        let object = NSManagedObject(entity: entity, insertIntoManagedObjectContext: backgroundContext)
        object.setValue("123", forKey: "remoteID")
        try! backgroundContext.save() // Here the changes will get merged with the main thread

        dispatch_async(dispatch_get_main_queue(), {
            let fetchRequest = NSFetchRequest(entityName: "User")
            fetchRequest.predicate = NSPredicate(format: "remoteID = 123")
            let user = try! self.dataStack.mainContext.executeFetchRequest(fetchRequest).first
            print(user)
        })
    }
}
```

**Comment by 3lvis** [comment]
I think I found your use case, this is what I have in my app, and two ways of doing it:

**Fetcher**:

```swift
func updateAlbum(id: String, title: String, description: String, completion: (error: NSError?) -> ()) {
    let parameters = [
        "title": title,
        "description": description
    ]
    self.networking.POST("/photos/v1/albums/\(id)", parameters: parameters) { JSON, error in
        self.dataStack.performInNewBackgroundContext({ backgroundContext in
            guard let JSON = JSON as? [String : AnyObject] else { fatalError("Album JSON has inconsistent structure") }

            let fetchRequest = NSFetchRequest(entityName: Album.entityName())
            fetchRequest.predicate = NSPredicate(format: "remoteID = %@", id)
            let album = try! backgroundContext.executeFetchRequest(fetchRequest).first as! Album
            album.hyp_fillWithDictionary(JSON)

            try! backgroundContext.save()

            dispatch_async(dispatch_get_main_queue(), {
                completion(error: error)
            })
        })
    }
}
```

**Controller A**:

```swift
func albumInfoController(albumInfoController: AlbumInfoController, didSaveWithTitle title: String, description: String) {
    self.fetcher.updateAlbum(self.album.remoteID!, title: title, description: description) { error in
        self.dismissViewControllerAnimated(true, completion: nil)
        self.album.managedObjectContext?.refreshObject(self.album, mergeChanges: true)
    }
}
```

**Controller B**:

```swift
func albumInfoController(albumInfoController: AlbumInfoController, didSaveWithTitle title: String, description: String) {
    self.fetcher.updateAlbum(self.album.remoteID!, title: title, description: description) { error in
        self.dismissViewControllerAnimated(true, completion: nil)

        let fetchRequest = NSFetchRequest(entityName: Album.entityName())
        fetchRequest.predicate = NSPredicate(format: "remoteID = %@", self.album.remoteID!)
        self.album = try! self.album.managedObjectContext!.executeFetchRequest(fetchRequest).first as! Album
    }
}
```

**Comment by antonholmquist** [comment]
Thanks, I will have a look at this

**Comment by 3lvis** [comment]
I guess we could re-fetch **all** the items on the main thread and call `refreshObject` but this might be an overkill since if you're using NSFetchedResultsController it does it for you. I guess you want only **one** item to be refreshed the same way that I showed in my example above. In that case it's easier for you to just refresh that item doing:

```swift
self.album.managedObjectContext?.refreshObject(self.album, mergeChanges: true)
```

Closing this for now. Thanks for your input. Let me know if there's anything else we can do for you :heart:
